File: Unity/AI_FSM/Assets/Script/FSM.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

[CreateAssetMenu(fileName = "FSM asset", menuName = "FSM/Create new FSM")]
public /*abstract*/ class FSM : ScriptableObject
{
    [field:SerializeField] public State StartingState { get; private set; } = null;
    [field:SerializeField] public FSMComponent Owner { get; private set; }

    [SerializeField] State currentState = null;

    public virtual void StartFSM(FSMComponent _owner)
    {
        Owner = _owner;
		SetNextState(StartingState);
    }

    public void SetNextState(State _state)
    {
        if (!_state)
            return;
		currentState = Instantiate<State>(_state);
        currentState.Enter(this);
    }

    public void UpdateFSM()
    {
        currentState?.Update();
    }

    public virtual void StopFSM()
    {
        currentState?.Exit();
        currentState = null;
    }
}

File: Unreal/IA_Cleaner/Source/IA_Cleaner/Private/FSM/FSM.cpp
// Fill out your copyright notice in the Description page of Project Settings.


#include "FSM/FSM.h"
#include "FSM/FSMComponent.h"

void UFSM::StartFSM(UFSMComponent* _owner)
{
	if (!_owner)
		return;
	owner = _owner;
	SetNextState(StartingState);
}

void UFSM::UpdateFSM()
{
	if (currentState)
		currentState->Update();
}

void UFSM::StopFSM()
{
	if (currentState)
		currentState->Exit();
	currentState = nullptr;
}

void UFSM::SetNextState(TSubclassOf<UBaseState> _state)
{
	if (!_state)
		return;
	currentState = NewObject<UBaseState>(this, _state);
	currentState->Enter(this);
}

TObjectPtr<AActor> UFSM::GetActor()
{
	return owner->GetOwner();
}
